File: src/evoscry/providers/google.py
"""Google HTML scraper."""

from __future__ import annotations

import re
from datetime import datetime
from urllib.parse import urlencode

from bs4 import BeautifulSoup

from evoscry.http_client import fetch_with_config

GOOGLE_SEARCH_URL = "https://www.google.com/search"

DATE_RANGE_MAP = {
    "day": "qdr:d",
    "week": "qdr:w",
    "month": "qdr:m",
    "year": "qdr:y",
}


async def search_google(
    query: str,
    max_results: int = 10,
    language: str = "en",
    date_range: str | None = None,
) -> list[dict]:
    """Scrape Google HTML search and return raw results."""
    params: dict[str, str] = {
        "q": query,
        "num": str(min(max_results + 5, 40)),
        "hl": language,
    }
    if date_range and date_range in DATE_RANGE_MAP:
        params["tbs"] = DATE_RANGE_MAP[date_range]

    url = f"{GOOGLE_SEARCH_URL}?{urlencode(params)}"
    resp = await fetch_with_config(url)

    if resp.status_code in (429, 503):
        raise RuntimeError(
            f"Google rate-limited (HTTP {resp.status_code}). "
            "Try increasing EVOSCRY_REQUEST_DELAY_MS or using a proxy."
        )
    if resp.status_code != 200:
        raise RuntimeError(f"Google returned HTTP {resp.status_code}")

    return _parse(resp.text, max_results)


def _parse(html: str, max_results: int) -> list[dict]:
    soup = BeautifulSoup(html, "lxml")
    results: list[dict] = []

    for el in soup.select("div.g"):
        if len(results) >= max_results:
            break

        # Skip ads
        if el.find_parent(class_="uEierd"):
            continue
        if el.find_parent(attrs={"data-text-ad": True}):
            continue

        link = el.select_one("a")
        title_el = el.select_one("h3")
        if not link or not title_el:
            continue

        href = link.get("href", "")
        title = title_el.get_text(strip=True)
        if not href.startswith("http"):
            continue

        # Extract snippet
        snippet = ""
        for sel in ["div[data-sncf]", "div.VwiC3b", "span.aCOpRe"]:
            found = el.select_one(sel)
            if found:
                snippet = found.get_text(strip=True)
                break
        if not snippet:
            all_text = el.get_text()
            idx = all_text.find(title)
            if idx >= 0:
                snippet = all_text[idx + len(title) :].strip()[:300]

        # Try to extract date from snippet
        published_date = None
        date_match = re.match(r"^(\w{3}\s+\d{1,2},\s+\d{4})\s*[—–\-]\s*", snippet)
        if date_match:
            try:
                d = datetime.strptime(date_match.group(1), "%b %d, %Y")
                published_date = d.strftime("%Y-%m-%d")
                snippet = snippet[date_match.end() :]
            except ValueError:
                pass

        if title and href:
            results.append({
                "title": title,
                "url": href,
                "snippet": snippet,
                "engine": "google",
                "published_date": published_date,
            })

    return results

File: src/providers/google.ts
import * as cheerio from "cheerio";
import type { RawSearchResult, SearchOptions } from "./types.js";
import { fetchWithConfig } from "./http.js";

const GOOGLE_SEARCH_URL = "https://www.google.com/search";

const DATE_RANGE_MAP: Record<string, string> = {
  day: "qdr:d",
  week: "qdr:w",
  month: "qdr:m",
  year: "qdr:y",
};

export async function searchGoogle(options: SearchOptions): Promise<RawSearchResult[]> {
  const params = new URLSearchParams({
    q: options.query,
    num: String(Math.min(options.maxResults + 5, 40)), // request a few extra to account for filtering
    hl: options.language,
  });

  if (options.dateRange && DATE_RANGE_MAP[options.dateRange]) {
    params.set("tbs", DATE_RANGE_MAP[options.dateRange]);
  }

  const url = `${GOOGLE_SEARCH_URL}?${params.toString()}`;
  const response = await fetchWithConfig(url);

  if (!response.ok) {
    if (response.status === 429 || response.status === 503) {
      throw new Error(`Google rate-limited (HTTP ${response.status}). Try increasing EVOSCRY_REQUEST_DELAY_MS or using a proxy.`);
    }
    throw new Error(`Google returned HTTP ${response.status}`);
  }

  const html = await response.text();
  return parseGoogleResults(html, options.maxResults);
}

function parseGoogleResults(html: string, maxResults: number): RawSearchResult[] {
  const $ = cheerio.load(html);
  const results: RawSearchResult[] = [];

  // Google organic results live in div.g containers
  $("div.g").each((_i, el) => {
    if (results.length >= maxResults) return false;

    // Skip ad results
    if ($(el).closest("div.uEierd").length > 0) return;
    if ($(el).closest("[data-text-ad]").length > 0) return;

    const linkEl = $(el).find("a").first();
    const titleEl = $(el).find("h3").first();

    const href = linkEl.attr("href") || "";
    const title = titleEl.text().trim();

    // Skip non-http links (Google internal pages, etc.)
    if (!href.startsWith("http")) return;

    // Extract snippet from various known selectors
    let snippet = "";
    const snippetSelectors = [
      "div[data-sncf]",
      "div.VwiC3b",
      'div[style="-webkit-line-clamp:2"]',
      "span.aCOpRe",
    ];
    for (const sel of snippetSelectors) {
      const text = $(el).find(sel).first().text().trim();
      if (text) {
        snippet = text;
        break;
      }
    }

    // Fallback: grab any text after the title that looks like a description
    if (!snippet) {
      const allText = $(el).text();
      const titleIdx = allText.indexOf(title);
      if (titleIdx >= 0) {
        snippet = allText.slice(titleIdx + title.length).trim().slice(0, 300);
      }
    }

    // Try to extract date from snippet
    let publishedDate: string | null = null;
    const dateMatch = snippet.match(/^(\w{3}\s+\d{1,2},\s+\d{4})\s*[—–-]\s*/);
    if (dateMatch) {
      try {
        const d = new Date(dateMatch[1]);
        if (!isNaN(d.getTime())) {
          publishedDate = d.toISOString().split("T")[0];
          snippet = snippet.slice(dateMatch[0].length);
        }
      } catch {
        // ignore invalid date
      }
    }

    if (title && href) {
      results.push({
        title,
        url: href,
        snippet,
        engine: "google",
        publishedDate,
      });
    }
  });

  return results;
}
